plugins/single_plugins/idle.cpp

```diff
@@ -4,69 +4,250 @@ extern "C"
+        cube_zoom_speed_updated();
+    }
+
+    void destroy_dpms_timeout()
+    {
+        if (timeout_dpms)
+        {
+            on_idle_dpms.disconnect();
+            on_resume_dpms.disconnect();
+            wlr_idle_timeout_destroy(timeout_dpms);
+        }
+
+        timeout_dpms = NULL;
     }
 
-    void destroy_timeout()
+    void destroy_screensaver_timeout()
     {
-        if (timeout)
+        if (screensaver_enabled)
+            stop_screensaver();
+
+        if (timeout_screensaver)
         {
-            on_idle.disconnect();
-            on_resume.disconnect();
-            wlr_idle_timeout_destroy(timeout);
+            on_idle_screensaver.disconnect();
+            on_resume_screensaver.disconnect();
+            wlr_idle_timeout_destroy(timeout_screensaver);
         }
 
-        timeout = NULL;
+        timeout_screensaver = NULL;
     }
 
-    void create_timeout(int timeout_sec)
+    void create_dpms_timeout(int timeout_sec)
     {
-        destroy_timeout();
+        destroy_dpms_timeout();
         if (timeout_sec <= 0)
             return;
 
-        timeout = wlr_idle_timeout_create(wf::get_core().protocols.idle,
+        timeout_dpms = wlr_idle_timeout_create(wf::get_core().protocols.idle,
             wf::get_core().get_current_seat(), 1000 * timeout_sec);
 
-        on_idle.set_callback([&] (void*) {
+        on_idle_dpms.set_callback([&] (void*) {
             set_state(wf::OUTPUT_IMAGE_SOURCE_SELF, wf::OUTPUT_IMAGE_SOURCE_DPMS);
         });
-        on_idle.connect(&timeout->events.idle);
+        on_idle_dpms.connect(&timeout_dpms->events.idle);
 
-        on_resume.set_callback([&] (void*) {
+        on_resume_dpms.set_callback([&] (void*) {
             set_state(wf::OUTPUT_IMAGE_SOURCE_DPMS, wf::OUTPUT_IMAGE_SOURCE_SELF);
         });
-        on_resume.connect(&timeout->events.resume);
+        on_resume_dpms.connect(&timeout_dpms->events.resume);
+    }
+
+    void create_screensaver_timeout(int timeout_sec)
+    {
+        destroy_screensaver_timeout();
+        if (timeout_sec <= 0)
+            return;
+
+        timeout_screensaver = wlr_idle_timeout_create(wf::get_core().protocols.idle,
+            wf::get_core().get_current_seat(), 1000 * timeout_sec);
+
+        on_idle_screensaver.set_callback([&] (void*) {
+            start_screensaver();
+        });
+        on_idle_screensaver.connect(&timeout_screensaver->events.idle);
+
+        on_resume_screensaver.set_callback([&] (void*) {
+            stop_screensaver();
+        });
+        on_resume_screensaver.connect(&timeout_screensaver->events.resume);
+    }
+
+    void screensaver_frame()
+    {
+        cube_control_signal data;
+        bool carried_out = true;
+
+        if (screensaver_stopping)
+            rotation = duration.progress(rot_transition);
+        else
+            rotation += cube_rotate_speed->as_double();
+        if (rotation >= M_PI * 2)
+            rotation = 0.0;
```

Comment:
Hm hm, are you sure this doesn't result in some glitches? Like, if the rotate speed is too big, you might get `rotation = 2*PI + 0.1` and then you jump back to `0` but we wanted `0.1`. Would be better to `rotation -= 2 * M_PI` if I understand the idea correctly.